libnetdata/ebpf/ebpf.c

```diff
@@ -522,3 +522,123 @@ void ebpf_update_module(ebpf_module_t *em)
 
     ebpf_update_module_using_config(em);
 }
+
+//----------------------------------------------------------------------------------------------------------------------
+
+/**
+ * Load Address
+ *
+ * Helper used to get address from /proc/kallsym
+ *
+ * @param fa address structure
+ * @param fd file descriptor loaded inside kernel.
+ */
+void ebpf_load_addresses(ebpf_addresses_t *fa, int fd)
+{
+    if (fa->addr)
+        return ;
+
+    procfile *ff = procfile_open("/proc/kallsyms", " \t:", PROCFILE_FLAG_DEFAULT);
+    if (!ff)
+        return;
+
+    ff = procfile_readall(ff);
+    if (!ff)
+        return;
+
+    fa->hash = simple_hash(fa->function);
+
+    size_t lines = procfile_lines(ff), l;
+    for(l = 0; l < lines ;l++) {
+        char *fcnt = procfile_lineword(ff, l, 2);
+        uint32_t hash = simple_hash(fcnt);
+        if (fa->hash == hash && !strcmp(fcnt, fa->function)) {
+            char addr[128];
+            snprintf(addr, 127, "0x%s", procfile_lineword(ff, l, 0));
+            fa->addr = (unsigned long) strtoul(addr, NULL, 16);
+            uint32_t key = 0;
+            bpf_map_update_elem(fd, &key, &fa->addr, BPF_ANY);
+        }
+    }
+
+    procfile_close(ff);
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+/**
+ * Set unique dimension
+ *
+ * Set one unique dimension for all vector position.
+ *
+ * @param algorithms the output vector
+ * @param length     number of elements of algorithms vector
+ * @param algortihm  algorithm used on charts.
+*/
+void ebpf_set_unique_dimension(int *algorithms, size_t length, int algorithm)
+{
+    size_t i;
+    for (i = 0; i < length; i++) {
+        algorithms[i] = algorithm;
+    }
+}
+
+/**
+ * Fill Histogram dimension
+ *
+ * Fill the histogram dimension with the specified ranges
+ */
+char **ebpf_fill_histogram_dimension(size_t maximum)
```

Comment:
This function doesn't work correctly. There are gaps in us range: 2-3, 4-5, 8-9...
![image](https://user-images.githubusercontent.com/22008543/120791213-9432d100-c53c-11eb-966e-bb955e4b017d.png)

Reply:
Unfortunately this looks like a problem on dashboard, please, take a look for the data when it is print on stdout:

```
CHART filesystem.ext4_read_latency '' 'ext4 latency for each read request.' 'calls/s' 'EXT4 latency (eBPF)' '' 'stacked' 2160 1
DIMENSION 0us->1us 0us->1us incremental 1 1
DIMENSION 1us->2us 1us->2us incremental 1 1
DIMENSION 3us->4us 3us->4us incremental 1 1
DIMENSION 5us->8us 5us->8us incremental 1 1
DIMENSION 9us->16us 9us->16us incremental 1 1
DIMENSION 17us->32us 17us->32us incremental 1 1
DIMENSION 33us->64us 33us->64us incremental 1 1
DIMENSION 65us->128us 65us->128us incremental 1 1
DIMENSION 129us->256us 129us->256us incremental 1 1
DIMENSION 257us->512us 257us->512us incremental 1 1
DIMENSION 513us->1ms 513us->1ms incremental 1 1
DIMENSION 1ms->2ms 1ms->2ms incremental 1 1
DIMENSION 3ms->4ms 3ms->4ms incremental 1 1
DIMENSION 5ms->8ms 5ms->8ms incremental 1 1
DIMENSION 9ms->16ms 9ms->16ms incremental 1 1
DIMENSION 17ms->32ms 17ms->32ms incremental 1 1
DIMENSION 33ms->65ms 33ms->65ms incremental 1 1
DIMENSION 66ms->131ms 66ms->131ms incremental 1 1
DIMENSION 132ms->262ms 132ms->262ms incremental 1 1
DIMENSION 263ms->524ms 263ms->524ms incremental 1 1
DIMENSION 525ms->1s 525ms->1s incremental 1 1
DIMENSION 1s->2s 1s->2s incremental 1 1
DIMENSION 3s->4s 3s->4s incremental 1 1
DIMENSION 5s->+Inf 5s->+Inf incremental 1 1
```

We have the dimensions set correctly, but when you called attention for this, I observed that it was possible to improve our algorithm to have more meaninfull dimensions  for milliseconds.